pkg/cleanup/controller/controller.go

```diff
@@ -0,0 +1,302 @@
+
+	c.broadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: c.client.CoreV1().Events(v1.NamespaceAll)})
+	defer c.broadcaster.Shutdown()
+
+	klog.Info("Waiting for informer caches to sync")
+	// Wait for the caches to be synced before starting workers
+	if ok := cache.WaitForCacheSync(ctx.Done(), c.nodeListerSynced, c.pvListerSynced, c.pvcListerSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	klog.Infof("Starting workers, count: %d", workers)
+	// Launch workers to process
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}
+
+	klog.Info("Started workers")
+
+	// Look for stale PVs and start timers for resource cleanup
+	c.startCleanupTimersIfNeeded()
+
+	<-ctx.Done()
+	klog.Info("Shutting down workers")
+
+	return nil
+}
+
+// runWorker is a long-running function that will continually call the
+// processNextWorkItem function in order to read and process a message on the entryQueue.
+func (c *CleanupController) runWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+// processNextWorkItem will read a single work item off the entryQueue and
+// attempt to process it, by calling the syncHandler.
+func (c *CleanupController) processNextWorkItem(ctx context.Context) bool {
+	key, shutdown := c.pvQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.pvQueue.Done(key)
+
+	pvName, ok := key.(string)
+	if !ok {
+		// Item is invalid so we can forget it.
+		c.pvQueue.Forget(key)
+		klog.Errorf("expected string in workqueue but got %+v", key)
+		return true
+	}
+
+	err := c.syncHandler(ctx, pvName)
+	if err != nil {
+		// An error occurred so re-add the item to the queue to work on later (has backoff to avoid
+		// hot-looping).
+		c.pvQueue.AddRateLimited(key)
+		klog.Errorf("error syncing %q: %w, requeuing", pvName, err)
+		return true
+	}
+
+	c.pvQueue.Forget(key)
+	return true
+}
+
+// syncHandler processes a PV by deleting the PVC bound to if it's
+// associated Node is gone.
+func (c *CleanupController) syncHandler(ctx context.Context, pvName string) error {
+	pv, err := c.pvLister.Get(pvName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// PV was deleted in the meantime, ignore.
+			klog.Infof("PV %q in queue no longer exists", pvName)
+			return nil
+		}
+		return err
+	}
+
+	nodeName, ok := common.NodeAttachedToLocalPV(pv)
+	if !ok {
+		// For whatever reason the PV isn't formatted properly so we will
+		// never be able to get its corresponding Node, so ignore.
+		klog.Errorf("error getting node attached to pv: %s", pv)
+		return nil
+	}
+
+	nodeExists, err := common.NodeExists(c.nodeLister, nodeName)
+	if err != nil {
+		return err
+	}
+	// Check that the node the PV/PVC reference is still deleted
+	if nodeExists {
+		return nil
+	}
+
+	pvClaimRef := pv.Spec.ClaimRef
+	if pvClaimRef == nil {
+		return nil
+	}
+
+	pvc, err := c.pvcLister.PersistentVolumeClaims(pvClaimRef.Namespace).Get(pvClaimRef.Name)
```

Comment:
I would also use UID from the `ClaimRef` to be sure it is the exact same object.

Reply:
Done